client/components/tinymce/plugins/media/plugin.jsx

```diff
@@ -234,32 +298,43 @@ function mediaButton( editor ) {
 				return;
 			}
 
-			// The `img` object can be a string or a DOM node. We need a
-			// normalized string to replace the post content markup
-			let imgString;
-			if ( img.outerHTML ) {
-				imgString = img.outerHTML;
+			let mediaString = '';
+
+			if ( mediaHasCaption ) {
+				if ( captionNode.outerHTML ) {
+					mediaString = captionNode.outerHTML;
+
+					captionNode.outerHTML = event.content;
+				} else {
+					mediaString = captionNode;
```

Comment:
Oh, actually this doesn't make sense.

So, initially, the `if` condition was used to differentiate between HTML mode and Visual mode. In Visual mode, we navigate between DOM nodes so we have to use `outerHTML` of a node.

However, in HTML mode, we use `REGEXP_IMG` to capture all the `img` tags so there is no `outerHTML` method on `img`.

In my case, when an image has caption shortcode and we are in HTML mode, the code will replace only the `<img>` tags in the string leaving out the `[caption]` shortcode.

This is a bug @artpi found before. Maybe we could leave it for another PR? Since then, I'll probably remove unneeded code.